libphal/phal_dump.C

```diff
@@ -169,473 +219,574 @@ void writeSbeData(const std::filesystem::path& dumpPath, uint32_t reasonCode,
 }
 
 /**
- * @brief Writes PIBMEM data to a file.
- * @param[in] dumpData A vector of uint64_t values representing the PIBMEM data
- * to be written.
- * @param[in] basePath The filesystem path of the file to which the data will be
- * written.
+ * @brief Write the dump file to the path
+ * @param[in] data Pointer to the data buffer
+ * @param[in] len Length of the buffer
+ * @param[in] dumpPath Path of the file
  *
- * @throw std::runtime_error Throws if there is an error opening the file or
- * during the file write operation.
+ * Throws FILE_OPERATION_FAILED in the case of error
  */
-void writePIBMEMDataToFile(const std::vector<uint64_t>& dumpData,
-			   const std::filesystem::path& basePath)
+void writeDumpFile(char* data, size_t len, std::filesystem::path& dumpPath)
 {
+	log(level::INFO, "writng dump data to file");
+	std::ofstream outfile{dumpPath, std::ios::out | std::ios::binary};
+	if (!outfile.good()) {
+		int err = errno;
+		log(level::ERROR,
+		    "Failed to open the dump file for writing err=%d", err);
+		throw dumpError_t(exception::FILE_OPERATION_FAILED);
+	}
+	outfile.exceptions(std::ifstream::failbit | std::ifstream::badbit |
+			   std::ifstream::eofbit);
 	try {
-		std::ofstream outfile{basePath,
-				      std::ios::out | std::ios::binary};
-		if (!outfile.good()) {
-			throw std::runtime_error(
-			    "Failed to open file for writing");
-		}
-		outfile.write(reinterpret_cast<const char*>(dumpData.data()),
-			      sizeof(uint64_t) * dumpData.size());
-	} catch (const std::exception& e) {
-		log(level::ERROR, "Error writing PIBMEM data to file: %s",
-		    e.what());
-		throw;
+		outfile.write(data, len);
+	} catch (const std::ofstream::failure& oe) {
+		int err = errno;
+		log(level::ERROR,
+		    "Failed to write to dump file err=%d error message=%s", err,
+		    oe.what());
+		throw dumpError_t(exception::FILE_OPERATION_FAILED);
 	}
+	outfile.close();
 }
 
 /**
- * @brief Collects and writes PIBMEM dump data for a given processor target.
- * @param[in] proc The processor target from which to collect the PIBMEM dump
- * data.
- * @param[in] dumpPath The filesystem path where the dump file will be written.
- * @param[in] baseFilename The base filename to use for the dump file. This name
- * will be used to form the complete file path along with the dumpPath.
+ * @brief Writes the dump data contents in the register into the dump file
  *
- * @throw std::runtime_error Throws if there is an error in collecting the
- * PIBMEM dump data or writing to the file.
+ * @tparam DumpRegVal A templated vector of various possible dump type registers
+ * @param dumpRegs Dump register
+ * @param basePath File base path
+ * @param dumpType Type of dump, e.g; pibms/pibmem/sbe local reg etc
  */
-void collectPIBMEMDump(struct pdbg_target* proc,
-		       const std::filesystem::path& dumpPath,
-		       const std::string& baseFilename)
+template <typename DumpRegVal>
+void writeDumpFileForDumpContents(std::vector<DumpRegVal>& dumpRegs,
```

Comment:
why this new function added?

Reply:
Rather than having four different functions for writing four different HWPs dump contents with same lines of code and logic I have tried to optimize it into a single function by making a templated one which will handle the dump contents from the all four HWPs.

Reply:
Again my point is if I am touching some sections of the existing code during a major enhancement process wouldn't it be better to get it tidy up as much as possible in one go? I am ready to discuss it further in a meeting perhaps.